src/main/cpp/src/mixed_sort_merge_join.cu

```diff
@@ -0,0 +1,759 @@
+filter_by_conditional_impl(rmm::device_uvector<cudf::size_type>&& left_indices,
+                           rmm::device_uvector<cudf::size_type>&& right_indices,
+                           cudf::table_device_view const& left_table,
+                           cudf::table_device_view const& right_table,
+                           cudf::ast::detail::expression_device_view device_expression_data,
+                           cudf::size_type num_intermediates,
+                           rmm::cuda_stream_view stream,
+                           rmm::device_async_resource_ref mr)
+{
+  auto const num_pairs = left_indices.size();
+  if (num_pairs == 0) {
+    return {std::make_unique<rmm::device_uvector<cudf::size_type>>(0, stream, mr),
+            std::make_unique<rmm::device_uvector<cudf::size_type>>(0, stream, mr)};
+  }
+
+  // Create a boolean mask for indices to keep
+  auto keep_mask = rmm::device_uvector<bool>(num_pairs, stream, mr);
+
+  // Launch kernel to evaluate expression for each pair with dynamic shared memory sizing
+  int current_device = 0;
+  CUDF_CUDA_TRY(cudaGetDevice(&current_device));
+  int max_shmem_per_block = 0;
+  CUDF_CUDA_TRY(cudaDeviceGetAttribute(
+    &max_shmem_per_block, cudaDevAttrMaxSharedMemoryPerBlock, current_device));
+
+  using intermediate_t = cudf::ast::detail::IntermediateDataType<has_nulls>;
+  auto const per_thread_bytes =
+    static_cast<int>(num_intermediates) * static_cast<int>(sizeof(intermediate_t));
+  int block_size = 256;  // default
+  if (per_thread_bytes > 0) {
+    int const max_by_shmem = max_shmem_per_block / per_thread_bytes;
+    if (max_by_shmem > 0) {
+      // Prefer multiples of warp size (32)
+      int const rounded = (max_by_shmem / 32) * 32;
+      block_size        = std::max(32, std::min(256, rounded > 0 ? rounded : max_by_shmem));
+    } else {
+      block_size = 32;  // minimal reasonable size
+    }
+  }
+  cudf::size_type const grid_size = (num_pairs + block_size - 1) / block_size;
+  auto const shmem_size = static_cast<size_t>(block_size) * static_cast<size_t>(per_thread_bytes);
+
+  filter_join_indices_kernel<has_nulls>
+    <<<grid_size, block_size, shmem_size, stream.value()>>>(left_indices.data(),
+                                                            right_indices.data(),
+                                                            num_pairs,
+                                                            left_table,
+                                                            right_table,
+                                                            device_expression_data,
+                                                            keep_mask.data());
+
+  // Surface any kernel launch errors immediately
+  CUDF_CUDA_TRY(cudaPeekAtLastError());
+
+  // Count the number of true values in the mask
+  auto const num_matches =
+    thrust::count(rmm::exec_policy(stream), keep_mask.begin(), keep_mask.end(), true);
```

Comment:
Seems like it would be pretty easy for the `filter_join_indices_kernel` to just compute and return this instead of invoking another kernel.